### ISSUE TITLE
Original game bug: AI builds simple ship line where one dock is in an obviously-landlocked lake

### ISSUE THREAD

**Issue by jgottula** [OpenRCT2/OpenLoco#51]
Just ran into this particular AI bug today while playing the game.

This AI company started up, and as its first action, built two docks (one by San Diego and another by Los Angeles) and placed a hovercraft at the San Diego dock.

Problem is... while the Los Angeles dock was toward the Pacific Ocean side, the San Diego dock was built in a small lake (geographically it appears it's meant to be the Salton Sea) that's completely disconnected from the Pacific Ocean or any other body of water. So the boat can't actually get from one dock to the other.

How hard could it be for the AI company logic to check that there's a passable, connected water route between the two docks in this maximally simplistic ship route? The game goes as far as to disallow land tool alterations in the water if the modification would impede active boat routes, so clearly it must have *some* statically-determined notion of the path the boat will take between its stops. But I guess there are some circumstances under which it just does a really poor job of pathfinding. Like, so poor that it believes the ship will take an overland route somehow.

Screenshots and save files below.

### Screenshot 1 of 4
Overview of the braindead ship route set up by the AI company.
![ai_ship_route_bug1](https://user-images.githubusercontent.com/822207/35370503-0efe1e56-0143-11e8-9b0b-964caa40a252.png)

### Screenshot 2 of 4
The AI company doesn't have a low Intelligence score or anything like that. (See game manual page 84 for description of Intelligence attribute.)
![ai_ship_route_bug2](https://user-images.githubusercontent.com/822207/35370562-616cd830-0143-11e8-8855-0757081e4a5e.png)

### Screenshot 3 of 4
Demonstration of water tiles near the San Diego dock where raising the land is disallowed. I raised land on all of the tiles where it was allowed; the one-tile-wide path where there's still water is the area that the game protects because it believes it to be part of the boat's current route.
![ai_ship_route_bug3](https://user-images.githubusercontent.com/822207/35370685-2d3f164e-0144-11e8-9367-9c13e079d62e.png)

### Screenshot 4 of 4
I also tried messing with the land in front of the Los Angeles dock; it actually allowed me to *completely* surround the dock with non-water tiles. This would never happen normally. Also, in the earlier screenshots you may have noticed that there's a narrow river a short distance inland that vaguely goes parallel to an imaginary line drawn between San Diego and Los Angeles; I was able to blockade that river with land tiles freely as well wherever I tried, with no complaint from the game.
![ai_ship_route_bug4](https://user-images.githubusercontent.com/822207/35370684-28f3ad7a-0144-11e8-9f57-62a6cb137c68.png)

## Save Files

I was making incremental saves with this particular game, so I have one from shortly before the AI company started up, as well as one from after it started up and built its docks in idiotic locations.

### Save File 1 of 2: May 11, 1980

[North America (West) 1980 13.zip](https://github.com/OpenRCT2/OpenLoco/files/1662590/North.America.West.1980.13.zip)

This is the latest save file I have from before the AI company starts up. I moved the camera over to the San Diego / Los Angeles area and re-saved for testing convenience, but otherwise left it exactly the same.

If you load up this save file, Rottenfish Transport will pop into existence on May ~23; he'll start "Surveying landscape near San Diego" on May ~30; and then he'll start building his ship route on June 1.

Annoyingly, with this save file, while he does build essentially the same ship route, he consistently builds the San Diego dock on the Pacific Ocean side, in exactly the same spot every time (i.e. he doesn't reproduce the buggy behavior where he builds it in the Salton Sea). I suspect this must be due to the scenario random seed being a bit different than it was when I was actually playing (originally, I was building railroad tracks over to the East at the time, which presumably influenced the random seed I guess).

### Save File 2 of 2: May 29, 1980

[North America (West) 1980 14.zip](https://github.com/OpenRCT2/OpenLoco/files/1662602/North.America.West.1980.14.zip)

This is the save file that I made *immediately* after the AI company started up and had built its docks, but before it placed the ship down at the San Diego dock. You'll notice the slightly different dock placement compared to the earlier save file, as well as the slightly earlier build date, due to different influences from the game's random system.

With this save file, you can enjoy watching the hovercraft go around and around in circles in the Salton Sea after it departs the San Diego dock; and you can try out some of the land modification stuff that I described earlier for yourself.

**Comment by jgottula** [comment]
Oh wow.

I continued playing the same save game (the version of it where the AI built the docks nonsensically, i.e. the second save file in the original post); and now, much later, I happened to go back over to that area and discovered that the AI company had noticed that its ship was stuck, and as a consequence, removed both the docks and the ship. (This is normal behavior when the AI discovers that one of its routes isn't working after a while.)

But you know what it did? It rebuilt effectively the same route (dock, dock, hovercraft), except this time, it built the Los Angeles dock on the inland side, and the San Diego dock on the ocean side! Still broken, just in the opposite manner! It's preposterous!

Here's the save file for that:
[North America (West) 1980 28.zip](https://github.com/OpenRCT2/OpenLoco/files/1662707/North.America.West.1980.28.zip)

The mind boggles...

**Comment by jgottula** [comment]
The idiocy continues! 😝

Same game. About a year later. Another transport company (a different one) started up, and built the *exact same* broken dock/dock/hovercraft route. One dock near LA on the ocean side; other dock near SD on the inland side.

No I am not making this up. Here's the save file, with the view focused on the purple dude's broken ship setup right after he built it:
[North America (West) 1980 37.zip](https://github.com/OpenRCT2/OpenLoco/files/1662854/North.America.West.1980.37.zip)

**Comment by janisozaur** [comment]
Wow.

Can you check if route gets re-calculated if you build a river between the docks? Does that allow the ships to pass? Can you then still remove the river or is it locked?

**Comment by IntelOrca** [comment]
> The game goes as far as to disallow land tool alterations in the water if the modification would impede active boat routes...

I believe this is done by just setting a flag on tiles that a vehicle goes on.

**Comment by jgottula** [comment]
> Can you check if route gets re-calculated if you build a river between the docks? Does that allow the ships to pass? Can you then still remove the river or is it locked?

I mean, you can load up the save file yourself and see. But I'll take a look.

Oh and by the way, this save game has been played 100% on vanilla Locomotion, in case there was any question that any of the tiny bits of reimplemented code we've done had anything to do with this.

**Comment by jgottula** [comment]
Okay so I did the build-a-manmade-river experiment. (Fortunately I had oodles of money laying around in this particular game, so I was able to do massively expensive amounts of land and water editing without finding/inventing a money cheat.)

Each time I added new water tiles leading from the lake in the direction of the oceanside dock, the ship would move its spin-around-in-a-circle routine to the edge of the new water tiles. And then when I eventually connected up the last set of water tiles to the ocean water tiles, it went on its way, completing the route, and was able to go back and forth normally from that point in time forward.

Oh and once the ship had made a successful pass through, the game wouldn't let me remove the water tiles. But I didn't remember to test removing a water tile that the ship was *about* to pass through *before* it did, to confirm/deny whether the "cannot remove" flag is set when a ship actually passes over a tile or if it's set when the ship *decides* to path through a tile.

Here's a screenshot showing the final successful pathing:
![screenshot68](https://user-images.githubusercontent.com/822207/35377883-b6c67940-0165-11e8-9ebc-34f3fd0360dc.png)

**Comment by IntelOrca** [comment]
> The game goes as far as to disallow land tool alterations in the water if the modification would impede active boat routes...

I believe this is done by just setting a flag on tiles that a vehicle goes on. This is why if you for example build an alternative route for even your own buses you still can't remove the original road they were using. I have removed this check already because it is damn annoying.

**Comment by jgottula** [comment]
@IntelOrca Right, I saw what you wrote. I just still don't know for sure whether the game actually sets the flag on the tiles that the vehicle *physically goes on as it does so*, or if it sets that flag on the tiles that the vehicle plans to go on when it decides which way it's going to go. Either case could still explain the alternative-route thing.

And yeah, it's damn annoying that even after a company has removed its stations from a road route and taken the vehicles off, the city-built road segments that they were *previously using* are still marked as in-use.

**Comment by jgottula** [comment]
> How did you remove the water tile?

In which case?

**Comment by IntelOrca** [comment]
@jgottula nevermind, conclusion is that AI does not do any path check before building the two end points.

**Comment by Derpercoaster** [comment]
I was playing in this scenario recently, and an AI company made a ship route in that exact same area. They did not, however, make an impossible one, as seen below:
![North America (West) 1980](https://user-images.githubusercontent.com/60452987/139781401-3a98b52c-68b7-4f65-9306-ea5b544207c8.png)

Same cities, same ship, but no problem. The AI doesn't always run into this issue, it seems.
